### PR TITLE
feat(background-media): add support for background videos

### PR DESCRIPTION
**Pull request by andy-blum**
### Related Ticket(s)

#7500 

### Description

PR adds a new property to the video-container → video-composite → video-player setup, `background-mode`. When `true`, Kaltura API passes along some additional `flashVars` that cause the video to autoplay, mute, and hide the player UI. The `background-media` component is re-worked from an extension of `DDSImage` to being a container that places a gradient overtop of slotted content, allowing authers to enter either a `<dds-image>` component or a `<dds-video-container>` component

### Changelog

**New**

- `dds-video-player-container` now has "background-mode" attribute that autoplays & mutes video as well as hides the player UI
- `dds-video-player-composite` has getter/setter methods to read and update user preference regarding video autoplay. This allows users to pause a background video and indicate to future pageloads to _not_ autoplay videos. If a user has not previously indicated a preference, the component looks for the OS prefers-reduced-motion setting

**Changed**

- `dds-background-media` supports using `dds-video-container` components
- Unit test snapshots

### CODE REVIEW

**Comment by andy-blum** [comment]
@IgnacioBecerra @jeffchew can you confirm that this approach is acceptable?

File: packages/web-components/src/components/background-media/__stories__/background-media.stories.ts [hunk 1]
```diff
@@ -25,17 +25,23 @@ const gradientDirections = {
...
+    </dds-background-media>
+  `;
+};
+
+export const WithVideo = ({ parameters }) => {
+  const { gradientDirection } = parameters?.props?.['dds-background-media'] ?? {};
+  return html`
+    <dds-background-media gradient-direction="${ifNonNull(gradientDirection)}" mobile-position="bottom">
+      <dds-video-player-container playing-mode="inline" video-id="1_9h94wo6b" background-mode="true"></dds-video-player-container>
     </dds-background-media>
```
[the remark this answers is not in the document]
Comment (proeung) [reply packages/web-components/src/components/background-media/__stories__/background-media.stories.ts:45]: > For a time I was wondering if there was a good way for the <dds-background-media> component to detect what was inside it and to update accordingly.

@andy-blum That's my preference as well but would love to hear from other engineers from the DDS side as well.

**Comment by andy-blum** [comment]
The e2e test error isn't happening for me locally:

## Github log:
Screen Shot 2021-11-04 at 10 25 15 AM

---

## Local log:
Screen Shot 2021-11-04 at 10 23 56 AM

**Comment by jeffchew** [comment]
@andy-blum `@carbon/icons` doesn't appear to be a dependency for `@carbon/ibmdotcom-web-components`, which is why you're seeing this as an integration test issue. Not sure why it's working for you locally though.

**Comment by andy-blum** [comment]
@jeffchew How can/should I use carbon icons within the web-components package?

**Comment by IgnacioBecerra** [comment]
@andy-blum Could be worth to try using `carbon-web-components/es/icons` like it's used in some of the other components: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/component-mixins/cta/cta.ts#L13

**Comment by jeffchew** [comment]
@andy-blum also noticing the downstream app integration test, seeing this error coming up:

image

File: packages/web-components/src/components/video-player/video-player-composite.ts [hunk 2]
```diff
@@ -42,7 +42,19 @@ class DDSVideoPlayerComposite extends HybridRenderMixin(HostListenerMixin(LitEle
    *
    * @internal
    */
-  _embedMedia?: (videoId: string) => Promise<any>;
+  _embedMedia?: (videoId: string, backgroundMode: boolean) => Promise<any>;
```
Comment (jeffchew) [inline packages/web-components/src/components/video-player/video-player-composite.ts:45]: @andy-blum looks like this is a new boolean added which is causing errors in the downstream app.

@annawen1 @andysherman2121 in these cases would there be a default value set if it's not defined?

**Comment by jeffchew** [comment]
@andy-blum looks like the snapshot also needs to be updated:

image

File: packages/web-components/src/components/video-player/video-player-composite.ts (hunk 2 above)
Comment (andy-blum) [reply packages/web-components/src/components/video-player/video-player-composite.ts]: @jeffchew there's a default value on the `backgroundMode` parameter, but it's defined in video-player-container. I've updated the typescript for this file to correctly mark `backgroundMode` as optional.

**Comment by andy-blum** [comment]
I have read the DCO document and I hereby sign the DCO.

**Comment by proeung** [comment]
I have read the DCO document and I hereby sign the DCO.

**Comment by andy-blum** [comment]
recheck

**Comment by jeffchew** [comment]
Thanks @andy-blum !

A few of the components using background media are now showing blank now for some reason: 
image

**Comment by andy-blum** [comment]
@jeffchew Got the image displaying again, but some of the changes I've made in preparation for #7148 aren't playing nice with Safari.
Screen Shot 2021-11-10 at 12 10 31 PM
Screen Shot 2021-11-10 at 12 12 29 PM

**Comment by jeffchew** [comment]
@andy-blum sounds like this is related to this issue:
https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/7573

**Comment by proeung** [comment]
@andy-blum Changes look good overall, however, the `README.stories.mdx` file should be updated with the markup for the video background option. Do you mind updating the doc so that adopters are aware of the different custom elements needed to construct this component?

```
<dds-background-media
        gradient-direction="${ifNonNull(gradientDirection)}"
        mobile-position="bottom"
        opacity="${ifNonNull(backgroundOpacity)}"
      >
  <dds-video-player-container playing-mode="inline" video-id="1_9h94wo6b" background-mode="true">
  </dds-video-player-container>
</dds-background-media>
```

**Review by proeung — APPROVED** [review]
LGTM! Thanks for addressing my feedback.

The background media component is now rendering fine in the Card section offset. 👍 

@IgnacioBecerra or @ariellalgilmore Can you review this PR and let us know if Andy's changes are good for merging?

**Review by IgnacioBecerra — APPROVED** [review]
All the changes and the new video variation look good to me, just gotta wait for QA review to merge this in.

@DragosRistici Can you take a look when you have the chance? 😄

**Review by ariellalgilmore — APPROVED** [review]
LGTM as well!

**Comment by DragosRistici** [comment]
Talked to Andy on Slack. What a wonderful job! Waiting on a small check and we are good to go. Thank you, Andy!

**Review by DragosRistici — APPROVED** [review]
LGTM! Thank you, @andy-blum!

**Pull request merged**